text/0036-client-network-variables.md

```diff
@@ -0,0 +1,142 @@
+# Client Network Variables
+[feature]: #feature
+
+- Start Date: `2021-12-12`
+- RFC PR: [#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/pull/0000)
+- SDK PR: [#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/pull/0000)
+
+# Summary
+[summary]: #summary
+
+Today, on the server may write to `NetworkVariables`.  This RFC introduces a bifurcation where in addition to the existing (server) NetworkVariables, we now have ClientNetworkVariables, which clients may write to.
+
+# Motivation
+[motivation]: #motivation
+
+To enable users to more simply communicate to the server (especially input) without having to use RPCs, this feature is proposed.  It is an extension of the existing `NetworkVariables` differing mainly in that clients can write to them.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Today, one adds a (server) `NetworkVariable` by declaring one in a `NetworkBehaviour` like so:
+
+```
+public class NetworkVariableTest : NetworkBehaviour
+{
+    public NetworkVariable<int> SomeSeverVar = new NetworkVariable<int>();
+
+    // ...
+
+    void Update()
+    {
+        if (NetworkManager.isServer)
+        {
+            SomeServerVar.Value = 123; // OK
+        }
+
+        if (NetworkManager.isClient)
+        {
+            // InvalidOperationException: Client can't write to NetworkVariables
+            // SomeServerVar.Value = 456; // ERROR
+        }
+
+    }
+}
+```
+
+Note, writing to the `SomeServerVar` in the server case succeeds (and is replicated to all clients), but fails with indicated exception in the comment.
+
+This RFC proposes a dedicated `ClientNetworkVariable` type that will work like this:
+
+```
+public class NetworkVariableTest : NetworkBehaviour
+{
+    public NetworkVariable<int> SomeSeverVar = new NetworkVariable<int>();
+    public ClientNetworkVariable<int> SomeClientVar = new ClientNetworkVariable<int>();
+
+    // ...
+
+    void Update()
+    {
+        if (NetworkManager.isServer)
+        {
+            SomeServerVar.Value = 123; // OK
+            SomeClientVar.Value = 456; // ERROR!
+        }
+
+        if (NetworkManager.isClient)
+        {
+            SomeServerVar.Value = 123; // ERROR!
+            SomeClientVar.Value = 456; // OK
+        }
+
+    }
+}
+```
+
+When the `SomeClientVar` code is executed...
+- The `SomeClientVar` for the client on which the code ran is immediately udpated
+- The client then communicates this variable update change to the server
+- The server then broadcasts this to the other clients
+- Note, unlike pre-1.0 versions of NGO when clients could write to network variables, the server can **not** write to client network variables.
+
+## Private Mode
+
+So that developers can choose to replicate the `ClientNetworkVariable` just on the server but **not** to the other clients, a `OwnerOnly` mode is avaiable.
```

Comment:
I'm wondering, couldn't this use the interest system instead of having its own system? This way, we'd be able to set custom replication rules (for example, only broadcast to team members or only broadcast to steam friends)

Reply:
Very interesting.  The thinking behind it was, ClientNetvars are a distinct, special type.  A CNV will not be snapshotted.  A client can never write to a regular NV.  A CNV can be private to the owning client.  We could, rather than make it its own type, have attributes on a regular NV, and then those attributes could be read / processed by the interest system.  I guess the spirit was to really keep CNV and NV as distinct types so as to not imply to the user that CNV's are treated by the SDK just like NV's (mainly, that they aren't snapshotted / predicted).  But I'm open to more feedback

Reply:
Ah interesting. Yes having attributes on a regular NV, with no new CNV class, read by interest management would indeed make this a cohesive solution
Will CNVs "never" be snapshotted? I thought Jeff had plans to allow this?